[SYCL][Docs] Add proposed event mode extension (#15704)
intel/llvm: sycl/doc/extensions/proposed/sycl_ext_intel_low_power_event.asciidoc

`event` created from the corresponding commands to do low-power synchronization.
If the backend is able to handle low-power events, calling `event::wait()` or
`event::wait_and_throw()` will cause the thread to yield and only do occasional
wake-ups to check the event progress.

```
namespace sycl::ext::intel::experimental {

struct low_power_event_key {
  using value_t = 
      oneapi::experimental::property_value<low_power_event_key>;
};

inline constexpr low_power_event_key::value_t low_power_event;

} // namespace sycl::ext::intel::experimental
```

=== New property usage example

As an example of how to use the new `low_power_event` property, see the
following code:

```
#include <sycl/sycl.hpp>

namespace oneapiex = sycl::ext::oneapi::experimental;
namespace intelex = sycl::ext::intel::experimental;

int main() {
  sycl::queue Q;

  // Submit some work to the queue.
  oneapiex::submit(Q, [&](sycl::handler &CGH) {...});

  // Submit a barrier with the low-power event property.
  sycl::event E = oneapiex::submit_with_event(Q, [&](sycl::handler &CGH) {
    oneapiex::barrier(CGH, oneapiex::properties{intelex::low_power_event});

>   });

Reviewer: I'm not a fan of putting the new property list in the command like this.  None of the other commands accept a property list, so `barrier` would be the exception.  We also have some commands that accept a parameter pack of reductions at the end, and redefining that as a parameter pack of reductions _or_ a final property list would get us back to the complexity we've been trying to avoid.

What about moving the property list to `submit_with_event`:
```c++
  // Submit a barrier with the low-power event property.
  sycl::event E = oneapiex::submit_with_event(Q, [&](sycl::handler &CGH) {
    oneapiex::barrier(CGH);
  }, oneapiex::properties{intelex::low_power_event});
```

This would have the advantage that a developer could create a low-powered event from _any_ command, not just barriers.

Author: > This would have the advantage that a developer could create a low-powered event from any command, not just barriers.

That is not currently feasible and one of the reasons why we restrict it to barrier. I think my main concern with tying it to the submit function is that users may get the wrong idea about where the property can be used. I suppose we might be able to throw an exception in the case that it is passed to a submit that does not resolve to a barrier command though.

Author: Okay, I think I agree with that point. I've changed it to add properties to `submit` and `submit_with_event` and changed the wording to say that the property is only valid with the latter. I've also added a note saying that it currently only has an effect on L0 submissions with barriers.

from the corresponding commands to do low-power synchronization. If the backend
is able to handle low-power events, calling `event::wait()` or
`event::wait_and_throw()` will cause the thread to yield and only do occasional
wake-ups to check the event progress.

[_Note:_ The property currently only has an effect on `barrier` and
`partial_barrier` commands enqueued on queues that return
`backend::ext_oneapi_level_zero` from `queue::get_backend()`.
_{endnote}_]

```
namespace sycl::ext::intel::experimental {

struct low_power_event_key {
  using value_t = 
      oneapi::experimental::property_value<low_power_event_key>;
};

inline constexpr low_power_event_key::value_t low_power_event;

} // namespace sycl::ext::intel::experimental
```

=== New property usage example

As an example of how to use the new `low_power_event` property, see the
following code:

```
#include <sycl/sycl.hpp>

namespace oneapiex = sycl::ext::oneapi::experimental;
namespace intelex = sycl::ext::intel::experimental;

int main() {
  sycl::queue Q;

  // Submit some work to the queue.
  oneapiex::submit(Q, [&](sycl::handler &CGH) {...});

  // Submit a command with the low-power event property.

>   oneapiex::properties Props{intelex::low_power_event};

Reviewer 2: Presently, our `wait` calls end up in the UR which calls `zeEventHostSynchronize` on L0.  ( https://spec.oneapi.io/level-zero/0.95/core/api.html#zeeventhostsynchronize ).  That call takes a nanoseconds arg, which is how long it block before reporting back not completed.  Presently the UR passes `UINT64_MAX` which makes it block fully until the operation completes.  Is there a different API call planned to be used?

My worry with this design is that we are handwaving and saying "do low power", but what is that, exactly? Personally, I think the user should provide the sleep timeout. That way they are in control of how much accuracy they want to sacrifice for how much less CPU usage.

Reviewer 2: nm, apparently the L0 call will behave differently for low power zeEvents.

:endnote: &#8212;{nbsp}end{nbsp}note

// Set the default source code type in this document to C++,
// for syntax highlighting purposes.  This is needed because
// docbook uses c++ and html5 uses cpp.
:language: {basebackend@docbook:c++:cpp}

:common_ref_sem: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:reference-semantics

== Notice

[%hardbreaks]
Copyright (C) 2024 Intel Corporation.  All rights reserved.

Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
permission by Khronos.

== Contact

To report problems with this extension, please open a new issue at:

https://github.com/intel/llvm/issues

== Dependencies

This extension is written against the SYCL 2020 revision 9 specification.  All
references below to the "core SYCL specification" or to section numbers in the
SYCL specification refer to that revision.

This extension also depends on the following other SYCL extensions:

* link:../experimental/sycl_ext_oneapi_enqueue_functions.asciidoc[
  sycl_ext_oneapi_enqueue_functions]
* link:../experimental/sycl_ext_oneapi_properties.asciidoc[

>   sycl_ext_oneapi_enqueue_properties]

Reviewer: ```suggestion
  sycl_ext_oneapi_properties]
```

== Status

This is a proposed extension specification, intended to gather community
feedback.  Interfaces defined in this specification may not be implemented yet
or may be in a preliminary state.  The specification itself may also change in
incompatible ways before it is finalized.  *Shipping software products should
not rely on APIs defined in this specification.*

== Backend support status

> This extension is currently implemented in {dpcpp} only for all device targets.

Reviewer 3: Are we just saying here that DPC++ is the only SYCL compiler that implements this extension?  I don't think we need to say that in the spec.  We can just remove the "Backend support status" section in that case.

== Overview

> On some backends, calling `wait()` on a `event` will synchronize using a

Reviewer 3: ```suggestion
On some backends, calling `wait()` on an `event` will synchronize using a
```

busy-waiting implementation. Though this comes at a low latency for the
synchronization of the event, it has the downside of consuming high amounts of
CPU time for no meaningful work.  This extension introduces a new property for
SYCL commands that will produce a "low-power" event. These new low-power events
will, if possible, yield the thread that the `wait()` member function is called
on and only wake up occasionally to check if the event has finished. This
reduces the time the CPU spends checking finish condition of the wait, at the
cost of latency.

== Specification

=== Feature test macro

This extension provides a feature-test macro as described in the core SYCL
specification.  An implementation supporting this extension must predefine the

> macro `SYCL_EXT_ONEAPI_LOW_POWER_EVENT` to one of the values defined in the table

Reviewer 4: Since this extension name doesn't have "oneapi" in the title, not sure if this should be?
```suggestion
macro `SYCL_EXT_INTEL_LOW_POWER_EVENT` to one of the values defined in the table
```

below.  Applications can test for the existence of this macro to determine if
the implementation supports this feature, or applications can test the macro's
value to determine which of the extension's features the implementation
supports.

[%header,cols="1,5"]
|===
|Value
|Description

|1
|The APIs of this experimental extension are not versioned, so the
 feature-test macro always has this value.
|===

=== Low-power event property

This extension adds a new property which can be used with the
`submit_with_event` free function from
link:../experimental/sycl_ext_oneapi_enqueue_functions.asciidoc[sycl_ext_oneapi_enqueue_functions].

> Passing this property to either of these functions will act as a hint to the

Reviewer 3: ```suggestion
Passing this property to this function will act as a hint to the
```

It seems like there is only one function, right?

`event` created from the corresponding commands to do low-power synchronization.
If the backend is able to handle low-power events, calling `event::wait()` or
`event::wait_and_throw()` will cause the thread to yield and only do occasional
wake-ups to check the event progress.

[_Note:_ The property currently only has an effect on `barrier` and
`partial_barrier` commands enqueued on queues that return
`backend::ext_oneapi_level_zero` from `queue::get_backend()`.
_{endnote}_]

```
namespace sycl::ext::intel::experimental {

struct low_power_event_key {
  using value_t = 
      oneapi::experimental::property_value<low_power_event_key>;
};

> inline constexpr low_power_event_key::value_t low_power_event;

Reviewer 3: I think this should be a runtime property that takes a `bool` parameter.  This will allow usage like:

```
bool low_power = /* check some option */;
oneapiex::properties props{intelex::low_power_event{low_power}});
```

Code like this is difficult with the API as you have it now because the existence of the `low_power_event` property affects the type of the `props` variable.  As a result, you can *not* do this:

```
bool low_power = /* check some option */;
oneapiex::properties props;
if (low_power)
  /* no way to add 'low_power_event' property now */
```

If you make it a runtime property, the property's default constructor can set the property's value to `true`.

Alternatively, the property could take an enum instead of a `bool`.  I think someone suggested in one of our meetings that there could be other types of low-power events in the future?  I can't remember whether we decided that was realistic, though.

Author: That's a good point. Would it make sense to change the property to an `event_mode` property and have the value as:

```
enum event_mode_enum {
  default_mode,
  low_power_mode
};
```

If we think future modes might overlap, we can also do it as a bitmap.

Reviewer: Am I right in thinking that we can define it as an `enum` now, and convert it to a bitmask later just by defining the relevant operators?

Either way, I like this direction.  I'd be tempted to change the names, though, to avoid some redundancy:
```c++
enum class event_mode {
  none, // we can't use"default" because it's a keyword
  low_power,
};
```

Author: > Am I right in thinking that we can define it as an `enum` now, and convert it to a bitmask later just by defining the relevant operators?

Absolutely. As long as it doesn't cross the library boundary, we can change it later if we want.

Reviewer 3: I think the enum direction makes sense, but I'm not sure what names you are proposing.  It seems like the name `event_mode` has been proposed as both the name of the property and the name of the enumeration.  Can you write out an example usage to illustrate the naming that you propose?

Author: I'm definitely not married to the name of the enum, but I just took inspiration from the naming of the compile-time property enums.

Adjusting the example in the text, the property would be used as
```c++
oneapiex::properties Props{intelex::event_mode{intelex::event_mode_enum::low_power}};
```
Note, I agree with @Pennycook in that it should be an enum class, mainly so it doesn't bleed its contents.

} // namespace sycl::ext::intel::experimental
```

=== New property usage example

As an example of how to use the new `low_power_event` property, see the
following code:

```
#include <sycl/sycl.hpp>

namespace oneapiex = sycl::ext::oneapi::experimental;
namespace intelex = sycl::ext::intel::experimental;

int main() {
  sycl::queue Q;

  // Submit some work to the queue.
  oneapiex::submit(Q, [&](sycl::handler &CGH) {...});

  // Submit a command with the low-power event property.
  sycl::event E = oneapiex::submit_with_event(Q, [&](sycl::handler &CGH) {
    ...

>   }, oneapiex::properties{intelex::low_power_event});

Reviewer 3: I think we had debates in the past about the position of the lambda parameter in calls like this.  I think we prefer to make the lambda last when possible because it is usually multiple lines.  If there are parameters after the lambda, they are far away from the name of the function (`submit_with_event`), so it becomes less clear that they are parameters so that function call.

Could we instead add new overloads of `submit_with_event` and `submit` which take the property list before the lambda?
